osf/models/subject.py

```diff
@@ -29,8 +36,32 @@ def child_count(self):
     def get_absolute_url(self):
         return self.absolute_api_v2_url
 
-    @property
+    @cached_property
     def hierarchy(self):
-        if self.parents.exists():
-            return self.parents.first().hierarchy + [self._id]
+        if self.parent:
+            return self.parent.hierarchy + [self._id]
         return [self._id]
+
+    @cached_property
+    def object_hierarchy(self):
+        if self.parent:
+            return self.parent.object_hierarchy + [self]
+        return [self]
+
+    @classmethod
+    def create(cls, text, provider, parent=None, bepress_subject=None):
+        validate_subject_hierarchy_length(parent)
+        validate_subject_provider_mapping(provider, bepress_subject)
+        subject = cls(text=text, provider=provider, parent=parent, bepress_subject=bepress_subject)
+        subject.save()
+        return subject
+
+    def save(self, *args, **kwargs):
+        if PreprintService.objects.filter(subjects=self).exists():
```

Comment:
Better to use a `related_name` rather than query on the model in order to use Django's query cache.

```
if self.preprint_services.exists():
```